Dependency Update: add support for Visual Studio 2022, update Boost to 1.78, improve wxWidgets patch

scottmudge opened the pull request:
While compiling OrcaSlicer with Visual Studio 2022, I was having a few issues. The batch file did not work for me, so I used the standard procedure (using CMake manually), but still encountered a few issues:

1. The current version of boost (1.75) does not support Visual Studio 2022. 
2. The CMake command used to compile wxWidgets applies a custom patch file to the repo to make changes required by BambuSlicer/OrcaSlicer. This patch-file approach is problematic, and the CMake script will throw errors if it attempts to re-patch the same cloned repo on subsequent dependency rebuilds.

I addressed point (1) by updating Boost to 1.78. There were essentially no compatibility issues, besides needing to switch `boost::to_string` to `std::to_string` in one location.

I addressed point (2) by forking wxWidgets, applying the patch to the target `v3.1.5` tag, and then saving that patched commit to a custom branch. I retagged this patched commit as `v3.1.5-bambu`. This will allow a much easier time updating wxWidgets, as instead of manually needing to update a patch file line-by-line, it will be possible to merge new versions of wxWidgets (at different tags) into the already patched branch.

It also addresses the issue of CMake throwing errors if attempting to rebuild dependencies without completely clearing the CMake cache, as it won't attempt to re-patch the same repo twice.

SoftFever commented on deps/Boost/Boost.cmake:
```diff
@@ -147,8 +147,8 @@ if (NOT IS_CROSS_COMPILE OR NOT APPLE OR BUILD_SHARED_LIBS)
     dep_Boost
-   URL "https://boostorg.jfrog.io/artifactory/main/release/1.75.0/source/boost_1_75_0.tar.gz"
-    URL_HASH SHA256=aeb26f80e80945e82ee93e5939baebdca47b9dee80a07d3144be1e1a6a66dd6a
+   URL "https://boostorg.jfrog.io/artifactory/main/release/1.78.0/source/boost_1_78_0.tar.gz"
```
Hehe, I had trouble cross-compiling OrcaSlicer for Mac, which was required for CICD since there is no M1 runner. That's why I kept the Boost library at 1.75

But I think we can upgrade it only for Windows first.
Can you make the change for Windows only?

scottmudge replied:
Sure thing!

SoftFever commented on deps/wxWidgets/wxWidgets.cmake:
```diff
@@ -28,9 +28,8 @@ endif ()
 
 bambustudio_add_cmake_project(wxWidgets
-    GIT_REPOSITORY "https://github.com/wxWidgets/wxWidgets"
+    GIT_REPOSITORY "https://github.com/scottmudge/wxWidgets-bambustudio"
```
I agree the patch is not an elegant solution.
Though I think we have issues here: different OS requires different patches :(
From the maintenance point of view, the patch is still easier.

scottmudge replied:
Ah, yeah there are some OS switches in the CMake script. Forgot about that. Alright, I'll keep this change local for now.

SoftFever commented:
Thanks @scottmudge 
Please see my comments.
I suggest we only do the 1.78 boost upgrade for Windows to support VS2022 in this PR.
What do you think?

scottmudge commented:
> Thanks @scottmudge Please see my comments. I suggest we only do the 1.78 boost upgrade for Windows to support VS2022 in this PR. What do you think?

Makes sense to me. As far as the patch issue goes, I'll look into modifying the CMake command so that it doesn't re-apply a patch (or resets/re-clones the repo every build) over the same local copy and result in a bunch of errors. Which was my main problem with it.

SoftFever commented:
> > Thanks @scottmudge Please see my comments. I suggest we only do the 1.78 boost upgrade for Windows to support VS2022 in this PR. What do you think?
> 
> Makes sense to me. As far as the patch issue goes, I'll look into modifying the CMake command so that it doesn't re-apply a patch (or resets/re-clones the repo every build) over the same local copy and result in a bunch of errors. Which was my main problem with it.

Take a look at this line in wxWidgets.cmake
With the power of bash on *nix system, it is possible.
Windows need some other trick to do it.
```
if (MSVC)
    set(_patch_cmd ${PATCH_CMD} ${CMAKE_CURRENT_LIST_DIR}/0001-wxWidget-fix.patch)
else ()
    set(_patch_cmd test -f WXWIDGETS_PATCHED || ${PATCH_CMD} ${CMAKE_CURRENT_LIST_DIR}/0001-wxWidget-fix.patch && touch WXWIDGETS_PATCHED)
endif ()
````

scottmudge commented:
> > > Thanks @scottmudge Please see my comments. I suggest we only do the 1.78 boost upgrade for Windows to support VS2022 in this PR. What do you think?
> > 
> > 
> > Makes sense to me. As far as the patch issue goes, I'll look into modifying the CMake command so that it doesn't re-apply a patch (or resets/re-clones the repo every build) over the same local copy and result in a bunch of errors. Which was my main problem with it.
> 
> Take a look at this line in wxWidgets.cmake With the power of bash on *nix system, it is possible. Windows need some other trick to do it.
> 
> ```
> if (MSVC)
>     set(_patch_cmd ${PATCH_CMD} ${CMAKE_CURRENT_LIST_DIR}/0001-wxWidget-fix.patch)
> else ()
>     set(_patch_cmd test -f WXWIDGETS_PATCHED || ${PATCH_CMD} ${CMAKE_CURRENT_LIST_DIR}/0001-wxWidget-fix.patch && touch WXWIDGETS_PATCHED)
> endif ()
> ```

Yeah that's what I'm working on right now, but Windows command prompt syntax/capabilities is pretty annoying. 

Another thing I don't understand is that cmake script completely overrides the `test -f` gate on the next line for Linux systems... what's the point of that?

```
if (CMAKE_SYSTEM_NAME STREQUAL "Linux")
    set(_patch_cmd ${PATCH_CMD} ${CMAKE_CURRENT_LIST_DIR}/0001-wxWidget-fix.patch)
endif ()
```

scottmudge commented:
I believe this should work for windows, but I'm testing now:

```
if (MSVC)
    set(_patch_cmd if not exist WXWIDGETS_PATCHED ( ${PATCH_CMD} ${CMAKE_CURRENT_LIST_DIR}/0001-wxWidget-fix.patch && type nul > WXWIDGETS_PATCHED ))
else ()
...
```

scottmudge commented:
Alright had to make a couple modifications. Replaced `${PATCH_CMD}` in the windows/MSVC patch command with its definition, as I had to wrap `${GIT_EXECUTABLE}` in quotes, so Windows wouldn't complain if the git executable path had spaces in it. But it all seems to work fine now.

I still don't understand why this line exists for Linux, but I didn't touch it:

```
if (CMAKE_SYSTEM_NAME STREQUAL "Linux")
    set(_patch_cmd ${PATCH_CMD} ${CMAKE_CURRENT_LIST_DIR}/0001-wxWidget-fix.patch)
endif ()
```

Other changes that were required -- Bambu seems to modify the Webview nuget package version in the wxWidgets patch file, but they never updated the associated hash. So MSVC would always complain (and fail to build) when the hash didn't match. So I updated the hash.

SoftFever commented:
Don't worry about the Apple builds's signature failures.

SoftFever commented:
Look all good.
Merged.

julie777 commented:
Is there an updated document somewhere describing how to build successfully?
I am still trying to make it work.

julie777 commented:
It took me forever to find this again.
Why is there a pull request, but no related issue?
Searching in issues for "Boost" did not find this.
